text/1015-full-text-search.rst

```diff
@@ -0,0 +1,479 @@
+specialized type for that: ``FTSResult``. This type is defined as follows::
+
+  type FTSResult {
+    property rank -> float64 {
+      default := 0;
+    }
+    property highlights -> array<str> {
+      default := [];
+    }
+  }
+
+If ranking was requested, then the value will be written in ``rank``.
+
+If highlighting the matched text was requested, then the appropriate value
+will be written in ``highlights`` array. The array is needed in case the
+matches are highlighted separately (e.g. only the matched words are returned,
+omitting the surrounding text).
+
+These enhanced FTS matches would be performed by the following built-in
+functions::
+
+  function fts_match(
+    doc: str,
+    query: ftsearch,
+    rank_opts: optional str = 'default',
+    highlight_opts: optional str = {}
+  ) -> FTSResult
+
+  function fts_match(
+    doc: Object,
+    query: ftsearch,
+    rank_opts: str = 'default',
+    highlight_opts: optional str = {}
+  ) -> FTSResult
+
+The ``rank_opts`` and ``highlight_opts`` can potentially be backend-specific,
+however either one of them can take the value of ``{}`` with the meaning that
+ranking/highlighting of the results is not necessary. Both of them should also
+always accept ``default`` as a valid argument indicating that
+ranking/highlighting of the results should be done in some basic manner
+(typically using some faster method).
+
+We also want some way of telling when one ``ftsearch`` value is encountered as
+part of another::
+
+  function contains(haystack: ftsearch, needle: ftsearch) -> bool
+
+This should return ``true`` if the lexemes of the "needle" query are contained
+in the "haystack", ignoring how exactly the lexemes are combined. This is most
+useful when the "needle" consists of a single lexeme and there's a need to
+check whether another query already contains it.
+
+
+Query DSL
+---------
+
+There's no way to encompass all the features of many different FTS
+implementations in a single clear search language spec. Instead we want to
+provide a simple search language to cover many common use-cases and to cover
+searches that are driven by user input.
+
+The DSL must be able to express the following:
+
+- all of the words must match (default)
+- any of the words must match
+- specific word must *not* be in the results
+- group together a part of the match (to combine rules)
+- escape a special character to be included as part of the match
+- match a phrase exactly
+- match a prefix (prefix + wildcard).
+
+There are three broad approaches to implementing ``ftsearch`` operations:
+
+1) The operations are encoded entirely in the ``str`` that's being cast or fed
+   into ``to_ftsearch``.
+
+   By default, punctuation and other symbols are ignored when performing FTS,
+   so we can safely use some symbols as part of the query language.
+
+2) Another option is to add operators to EdgeQL for the purpose of
+   manipulating the ``ftsearch`` values.
```

Comment:
I think this is very "dangerous" thing to do. If people will start using query builder to compose edgedb queries based on user queries, this can quickly become explosion of the query cache (both on EdgeDB and client side).